### PR TITLE
html-language-features: text document provider support for customData.html

### PR DESCRIPTION
**Pull request by sijakret**
This PR implements #135459
It adds the ability for for html-language-features extension to receive custom-data via [Text Document Provider](https://code.visualstudio.com/api/extension-guides/virtual-documents#textdocumentcontentprovider) and update on changes to these documents.
With the change extensions can provide contribute dynamic data via `customData.html`;

### CODE REVIEW

File: extensions/html-language-features/client/src/customData.ts [hunk 1]
```diff
@@ -26,9 +26,16 @@ export function getCustomDataSource(toDispose: Disposable[]) {
...
+	toDispose.push(workspace.onDidChangeTextDocument(e => {
+		const path = e.document.uri.toString();
+		if (pathsInExtensions.has(path) || pathsInWorkspace.has(path)) {
+			onChange.fire();
+		}
+	}));
+
 	return {
 		get uris() {
-			return pathsInWorkspace.concat(pathsInExtensions);
```
Comment (aeschli) [inline extensions/html-language-features/client/src/customData.ts:31]: what's the benefit of that change?

**Review by aeschli — CHANGES_REQUESTED** [review]

File: extensions/html-language-features/client/src/customData.ts (hunk 1 above)
Comment (sijakret) [reply extensions/html-language-features/client/src/customData.ts:31]: I converted the path path lists in to Set<string> so the check in line 31 becomes as fast as possible.
With the change, checking of a path has changed should be O(1) but returning the list has higher complexity.
If we want to have a `Set` to track the paths and avoid having to flatten it to an array in the getter `get uris()` we would need to maintain both a set and an array in parallel. One for filtering paths and one to return it. I can change that if you want to. If it makes sense also depends a bit on how often the getter is invoked, but I guess we don't know that..

Comment (aeschli) [reply extensions/html-language-features/client/src/customData.ts:31]: Ah, now I get it, thanks!

File: extensions/html-language-features/client/src/requests.ts [hunk 2]
```diff
@@ -20,34 +22,34 @@ export namespace FsReadDirRequest {
 
 export function serveFileSystemRequests(client: CommonLanguageClient, runtime: Runtime, subscriptions: { dispose(): any }[]) {
 	subscriptions.push(client.onRequest(FsContentRequest.type, (param: { uri: string; encoding?: string; }) => {
-		const uri = Uri.parse(param.uri);
-		if (uri.scheme === 'file') {
+		const uri = param.uri.match(uriScheme);
```
Comment (aeschli) [inline extensions/html-language-features/client/src/requests.ts:25]: why did you change this?

File: extensions/html-language-features/client/src/customData.ts [hunk 3]
```diff
@@ -94,12 +94,11 @@ function getCustomDataPathsFromAllExtensions(): Set<string> {
 		const customData = extension.packageJSON?.contributes?.html?.customData;
 		if (Array.isArray(customData)) {
 			for (const rp of customData) {
-				const uri = Uri.parse(rp);
-				if (uri.scheme === 'file') {
-					// only resolve file paths relative to extension
+				if (!uriScheme.test(rp)) {
+					// no schame -> resolve relative to extension
```
Comment (aeschli) [inline extensions/html-language-features/client/src/customData.ts:98]: typo on 'scheme'

**Review by aeschli — CHANGES_REQUESTED** [review]

File: extensions/html-language-features/client/src/requests.ts (hunk 2 above)
Comment (sijakret) [reply extensions/html-language-features/client/src/requests.ts:25]: Based on earlier comments I thought `Uri.parse` shall explicitly NOT be used to check for schemes.
I actually actually liked it better before but it feels a bit inconsistent to use sometimes use a bare regex and sometimes use `Uri.parse`. Don't have a string opinion here. I can change it back. Lat me know what you prefer.

One more important thing:
The `serveFileSystemRequests(..)` helper was already in the code, I did not create it.
However, it was unused - so someone had written that helper but never activated it (or deactivated it for some reason).
I just activated added in [client.onReady()](https://github.com/microsoft/vscode/pull/137557/files#diff-326f44a4671f469344f799a8c570e485b446e6b32fc74c9265bc2a5662295fe2R123) in htmlClient.ts (same way it is done in css-language-features where pretty much the same code has been active all along).
Was there a reason why these handlers were added but commented out? Seems a bit suspicious..

Comment (aeschli) [reply extensions/html-language-features/client/src/requests.ts:25]: uti is always in the URI format. So Uri.parse is ok to use.

The point is to no use `Uri.parse` on a relative path.

**Comment by aeschli** [comment]
I made some polish changes (and reverted the requestService change), so this looks all good now.

The missing `serveFileSystemRequests` was a bug and all that is unfinished work of myself. I'll work on that when after the PR is merged.

**Comment by aeschli** [comment]
@sijakret Thanks for all the work!

**Comment by sijakret** [comment]
@aeschli something got lost! https://github.com/sijakret/vscode/commit/f1455eabedfea1ae02bd68351b1c910896551371#diff-35b696cf6cdc005ac94347afd4160dbae29c3a336d5ce690501bbd67ec6fddbcL35
